xmtp_mls/src/identity/xmtp_id/identity.rs

```diff
@@ -71,112 +77,127 @@ impl Identity {
-        let builder = SignatureRequestBuilder::new(inbox_id.clone());
-        let mut signature_request = builder
-            .create_inbox(member_identifier.clone(), nonce)
-            .add_association(installation_public_key.to_vec().into(), member_identifier)
-            .build();
+            signature_request
+                .add_signature(Box::new(
+                    sign_with_installation_key(
+                        signature_request.signature_text(),
+                        sized_installation_key(signature_keys.private())?,
+                    )
+                    .await?,
+                ))
+                .await?;
+            signature_request
+                .add_signature(Box::new(
+                    sign_with_legacy_key(
+                        signature_request.signature_text(),
+                        legacy_signed_private_key,
+                    )
+                    .await?,
+                ))
+                .await?;
+            let identity_update = signature_request.build_identity_update()?;
+            api_client.publish_identity_update(identity_update).await?;
 
-        // We can pre-sign the request with an installation key signature, since we have access to the key
-        signature_request
-            .add_signature(Box::new(
-                sign_with_installation_key(
-                    signature_request.signature_text(),
-                    sized_installation_key(signature_keys.private())?,
-                )
-                .await?,
-            ))
-            .await?;
+            let identity = Self {
+                inbox_id: inbox_id.clone(),
+                installation_keys: signature_keys,
+                credential: create_credential(inbox_id)?,
+                signature_request: None,
+            };
 
-        signature_request
-            .add_signature(Box::new(
-                sign_with_legacy_key(
-                    signature_request.signature_text(),
-                    legacy_signed_private_key,
-                )
-                .await?,
-            ))
-            .await?;
+            Ok(identity)
+        } else {
+            let nonce = rand::random::<u64>();
+            let inbox_id = generate_inbox_id(&address, &nonce);
+            let mut builder = SignatureRequestBuilder::new(inbox_id.clone());
+            builder = builder.create_inbox(member_identifier.clone(), nonce);
 
-        let identity_update = signature_request.build_identity_update()?;
+            let mut signature_request = builder
+                .add_association(installation_public_key.to_vec().into(), member_identifier)
+                .build();
 
-        api_client.publish_identity_update(identity_update).await?;
+            // We can pre-sign the request with an installation key signature, since we have access to the key
+            signature_request
+                .add_signature(Box::new(
+                    sign_with_installation_key(
```

Comment:
It's a bit confusing signing with installation key in libxmtp today. OpenMls offers its own signing procedure with BasicCredential that is currently used in v3 (where any additional signing complexity like a prehash is abstracted), in identity we are using our own signing procedure with dalek, and then again ethers offers another signing procedure similar to dalek but more functions to call.

I'm wondering if it's worth abstracting these signing methods into a central place, with conversions between differnt types. Like an `InstallationKey` with just a `sign` method. I could convert a [u8; 32] to this, a INstallatonKeyProto, and a BasicCredential to this one struct. This could make signing code more concise and easier to follow, as well as potentially reduce any errors we get by screwing up the prehash stuff.

I think a standalone function is OK for now, but i worry about the discoverability of it, esp as it's tucked away deep in an identity module

Reply:
I totally agree, we could have a central place to handle cryptography / primitives conversion / utils so that other parts of code won't need to deal with the complexity.
It would be a fairly big clean up task, hope we get to prioritize that soon.

Reply:
➕